mlflow/data/dataframe_dataset.py

```diff
@@ -0,0 +1,372 @@
+    )
+
+
+def infer_mlflow_dtype(
+    dtype: nw.dtypes.DType,
+    col_name: str,
+    *,
+    allow_unknown: bool,
+) -> ColSpecType:
+    dtype_cls = type(dtype)
+    mapped = EXACT_DTYPE_MAPPING.get(dtype_cls)
+    if mapped is not None:
+        return mapped
+
+    mapped = APPROX_DTYPE_MAPPING.get(dtype_cls)
+    if mapped is not None:
+        logging.warning(
+            "Data type of Column '%s' contains dtype=%s which will be mapped to %s."
+            " This is not an exact match but is close enough",
+            col_name,
+            dtype,
+            mapped,
+        )
+        return mapped
+
+    if isinstance(dtype, (nw.Array, nw.List)):
+        return Array(infer_mlflow_dtype(dtype.inner, f"{col_name}.[]", allow_unknown=allow_unknown))
+
+    if isinstance(dtype, nw.Struct):
+        return Object(
+            [
+                Property(
+                    name=field.name,
+                    dtype=infer_mlflow_dtype(
+                        field.dtype, f"{col_name}.{field.name}", allow_unknown=allow_unknown
+                    ),
+                )
+                for field in dtype.fields
+            ]
+        )
+
+    return _handle_unknown_dtype(dtype=dtype, col_name=col_name, allow_unknown=allow_unknown)
+
+
+def _handle_unknown_dtype(dtype: Any, col_name: str, *, allow_unknown: bool) -> str:
+    if not allow_unknown:
+        _raise_unknown_type(dtype)
+
+    logging.warning(
+        "Data type of Columns '%s' contains dtype=%s, which cannot be mapped to any DataType",
+        col_name,
+        dtype,
+    )
+    return str(dtype)
+
+
+def _raise_unknown_type(dtype: Any) -> None:
+    msg = f"Unknown type: {dtype!r}"
+    raise ValueError(msg)
+
+
+class DataFrameDataset(Dataset, PyFuncConvertibleDatasetMixin, Generic[IntoDataFrameT]):
+    """
+    Represents a (eager) DataFrame for use with MLflow Tracking.
+    """
+
+    backend_name: str
+
+    def __init__(
+        self,
+        df: IntoDataFrameT,
+        source: DatasetSource,
+        targets: Optional[str] = None,
+        name: Optional[str] = None,
+        digest: Optional[str] = None,
+        predictions: Optional[str] = None,
+    ) -> None:
+        """
+        Args:
+            df: A eager DataFrame (such as pandas, polars or pyarrow).
+            source: The source of the pandas DataFrame.
+            targets: The name of the target column. Optional.
+            name: The name of the dataset. E.g. "wiki_train". If unspecified, a name is
+                automatically generated.
+            digest: The digest (hash, fingerprint) of the dataset. If unspecified, a digest
+                is automatically computed.
+            predictions: Optional. The name of the column containing model predictions,
+                if the dataset contains model predictions. If specified, this column
+                must be present in the dataframe (``df``).
+        """
+        nw_frame = nw.from_native(df, eager_only=True)
```

Comment:
We might want to add a few flags here:

- `pass_through=False`
- `allow_series=False`